File: ruby/saddle-points/saddle_points.rb
class Matrix
  attr_reader :rows, :columns, :saddle_points

  def initialize(str)
    @rows = str.each_line.map { |line| line.split.map(&:to_i) }

    @columns = @rows.transpose

    @saddle_points = @rows.each_with_index.with_object([]) do |(row, i), arr|
      row_max = row.max

      row.each_with_index do |elem, j|
        if elem == row_max
          if elem == @columns[j].min
            arr << [i, j]
          else
            break
          end
        end
      end
    end
  end
end

File: javascript/saddle-points/saddle-points.js
export const saddlePoints = (rows) => {
  let points = [];

  let columns = rows.transpose();

  rows.forEach(function(row, i) {
    let row_max = Math.max(...row);

    row.forEach(function(elem, j) {
      if (elem == row_max) {
        let col_min = Math.min(...columns[j]);

        if (elem == col_min)
          points.push({row: i + 1, column: j + 1});
        else
          return;
      }
    });
  });

  return points;
};

Array.prototype.transpose = function() {
  let width = this[0].length;

  let output = new Array(width);

  for (let i = 0; i < width; ++i)
    output[i] = this.map(row => row[i]);

  return output;
}
